### PR TITLE
internal/v1: support parent agents

### PR DESCRIPTION
**Pull request by mhilton**

### CODE REVIEW

File: internal/v1/users.go [hunk 1]
```diff
@@ -142,21 +142,26 @@ func (h *handler) CreateAgent(p httprequest.Params, u *params.CreateAgentRequest
...
-		Name:       u.FullName,
-		Groups:     u.Groups,
-		PublicKeys: pks,
-		Owner:      owner.ProviderID,
+		Username:     agentName + "@candid",
+		ProviderID:   store.MakeProviderIdentity("idm", agentName),
+		Name:         u.FullName,
+		Groups:       u.Groups,
+		PublicKeys:   pks,
+		ProviderInfo: map[string][]string{"creator": {string(owner.ProviderID)}},
```
Comment (rogpeppe) [inline internal/v1/users.go]: Maybe put the "creator" on a new line so it's more obvious?
(I missed it when glancing through)

File: internal/v1/users_test.go [hunk 2]
```diff
@@ -218,6 +218,27 @@ var setUserTests = []struct {
 	},
 }}
 
+func (s *usersSuite) TestSetUser(c *gc.C) {
+	c.Skip("deprecated")
```
Comment (rogpeppe) [inline internal/v1/users_test.go]: Is it worth keeping around if we're just going to skip it?

File: internal/v1/users_test.go [hunk 3]
```diff
@@ -344,25 +365,142 @@ func (s *usersSuite) TestCreateAgentWithGroups(c *gc.C) {
...
+
+	client, err := candidclient.New(candidclient.NewParams{
+		BaseURL: s.URL,
+		Client: &httpbakery.Client{
+			Client: httpbakery.NewHTTPClient(),
+			Key:    privKey1,
+		},
+		AgentUsername: string(resp.Username),
+	})
+	_, err = client.CreateAgent(s.Ctx, &params.CreateAgentRequest{
```
Comment (rogpeppe) [inline internal/v1/users_test.go]: c.Assert(err, gc.Equals, nil)

File: internal/v1/users.go [hunk 4]
```diff
@@ -127,7 +127,7 @@ func (h *handler) CreateAgent(p httprequest.Params, u *params.CreateAgentRequest
 	if err != nil {
 		return nil, errgo.Notef(err, "cannot find identity for authenticated user")
 	}
-	if owner.ProviderID.Provider() == "idm" && owner.Owner != "" {
+	if owner.ProviderID.Provider() == "idm" && owner.Owner != "" && !u.Parent {
 		// The authenticated user is an agent, so we don't allow it to create other agents.
```
Comment (albertodonato) [inline internal/v1/users.go]: this comment probably needs updating too

File: internal/v1/users.go [hunk 5]
```diff
@@ -127,14 +127,18 @@ func (h *handler) CreateAgent(p httprequest.Params, u *params.CreateAgentRequest
...
+		// allowed to create their own agents.
+		// TODO a nicer way to do this check might be to express
+		// it as a group permission - all non-agent users are in
+		// the "can create agents" group.
+		// TODO In the future, we might allow agents to create
+		// other agents, but we'll have to work out what to do
+		// about hierarchy - if agent A creates agent B, then A
+		// is removed from a group but its owner is still a
+		// member of that group, should B still have access to
+		// the group?
```
Comment (albertodonato) [inline internal/v1/users.go]: Are the TODOs here still accurate?

**Pull request merged**